src/x11/x.c

```diff
@@ -584,14 +584,56 @@ static dimension_t x_render_layout(cairo_t *c, colored_layout *cl, colored_layou
         int bg_height = MAX(settings.notification_height, (2 * settings.padding) + h);
         double bg_half_height = settings.notification_height/2.0;
         int pango_offset = (int) floor(h/2.0);
+        const float degrees = M_PI / 180.0;
 
         if (first) bg_height += settings.frame_width;
         if (last) bg_height += settings.frame_width;
         else bg_height += settings.separator_height;
 
-        cairo_set_source_rgb(c, cl->frame.r, cl->frame.g, cl->frame.b);
-        cairo_rectangle(c, bg_x, bg_y, bg_width, bg_height);
-        cairo_fill(c);
+        XWindowAttributes win_attr;
+        XGetWindowAttributes(xctx.dpy, xctx.win, &win_attr);
+
+        int width = win_attr.width + win_attr.border_width;
+        int height = win_attr.height + win_attr.border_width;
+
+        int min_width_height = width < height ? width : height;
+        int half_min_width_height = (int) floor(min_width_height / 2.0);
+
+        settings.corner_radius = MIN(half_min_width_height, settings.corner_radius);
+
+        x_win_round_corners();
```

Comment:
This function call sounds contrary to my understanding of this method. You create/render the layout in here, and then round the whole window (for every notification). But you call it also at the end of `x_win_draw`, where it makes IMHO much more sense.